Reduce build size + some fixes

TD-er opened the pull request.

TD-er committed:
[Minimal OTA] Exclude web pages from minimal OTA build

Exclude:
- Favicon
- Log viewer  (also JSON log)
- I2C scanner
- Timing Stats
- System Variables display.

TD-er committed:
[Modbus RTU] Move strings to flash

TD-er committed:
[Build] Fix uninitialized warning in Scheduler

TD-er committed:
[Minimal OTA] Exclude Github Copy button on minimal build

TD-er committed:
[WiFi] Fix WiFi reissue when gotIp event is delayed

A simple typo could result in unexpected failed reconnect attempts.

TD-er committed:
[Minimal OTA] Add defines to enable/disable any web page

TD-er committed:
[Minimal OTA]  Disable buttons for not included pages

TD-er committed:
[FEATURE_SD] Fix build issues when FEATURE_SD is defined (not finished)

When enabling the FEATURE_SD, something with the define of __FlashStringHelper is not correct anymore.

This needs a lot of changes which were needed anyway.
But also arrays of __FlashStringHelper must be changed.

All done so far, only need to have a look at    
static const __FlashStringHelper *gpMenu[8][3] = {

TD-er committed:
Fix JSON service stopped working (#2697)

TD-er commented:
@clumsy-stefan 
Ah, now that I have a PR created from it, I can actually read your comments a lot better now :)

TD-er committed:
[Cleanup] Fix issue ControllerSettings & extra checks TaskIndex

TD-er committed:
[clangtidy] Use sizeof(*this) instead of sizeof(this)

this returns a pointer while we need to know the object size.

TD-er committed:
[clangtidy] Clean up Commands/*

Const reference parameters instead of deepcopy and some other fixes.

TD-er committed:
[clangtidy] Clean up DataStructs/*

Mainly use of macro parenthesis

TD-er committed:
[clangtidy] Missing header guard defines

TD-er committed:
[clangtidy] Const reference function parameters

TD-er committed:
[clangtidy] Missing forward declaration of toString(float,int)

This was actually a bug, since only the toString(bool) was forward declared.
Meaning the conversion of a float to a string was not performed as expected when called from a .cpp file instead of a .ino file.

clumsy-stefan commented:
good ;) again, great work, I can use my 1M units again!! Wohoooo!

The new commits slightly increases minimal build size for me (~2k).

compiling without `USE_MQTT` fails as `src/Commands/MQTT.cpp` should be casted with `#ifdef USES_MQTT` at the top.

Also my personal view would be to also exclude `WEBSERVER_PINSTATES`, `WEBSERVER_SETUP` and `WEBSERVER_WIFI_SCANNER` in minimal builds...

in `define_plugin_sets.h` you check for `WEBSERVER_CUSTOM_BUILD_DEFINED` before actually defining the included parts. Then I think further below on the "minimal OTA Build" part you should also check for this and not just `#undef` a bunch of them...

all again just IMHO.....

clumsy-stefan commented:
in `ESPEasy_fdwdecl.h` from line 158 should be excluded from ESP32 builds
```
#if defined(ESP8266)
# include <ESP8266WiFiType.h>
#endif // defined(ESP8266)
```
or include it at the top (lines 14ff).

~~However compiling still fails on ESP32 as `WIFI_MODE_MAX` does not seem to be declared coreectly. It did compile before splitting it into sepparate `.cpp` and `.h` files. but I can't find why...~~
EDIT: seems to be a local issue....

TD-er commented:
> good ;) again, great work, I can use my 1M units again!! Wohoooo!

See also this one I added last night: https://github.com/esp8266/Arduino/issues/6691
So more free space at the horizon.

clumsy-stefan commented:
that would explain the significant increase after going to 2.6... 

Everything works smoothly on the new builds, expect I'm not able to compile for ESP32... did you already try?

TD-er commented:
@esp32
Not yet, but I noticed the Travis build also failed, so I will have a look at it.

clumsy-stefan commented:
seems like a missing `#endif` or similar, but can't find it...

TD-er commented:
Nope, some forward declaration I had to make, which needed an include.
My fault.

clumsy-stefan commented:
ah, ok, great, thanks for looking into it..

TD-er commented:
By the way, did you see al commits I made last night (literally) after using the static code analyzer?

It did find some true bugs.

clumsy-stefan commented:
yep, I did new versions for all my nodes this morning, including the whole PR here (all commits from your reduce_build_size_branch). I quickly browsed through the changes, and indeed seems like you found some things that could have caused problems... like pointers instead of values etc.... 

nice work... 

already updated all my nodes (yeah, I know a bit daring, but that's ok ;) and everything runs smoothly until now! Especially the 1M nodes work again... that's great!!

TD-er commented:
Do you use the `publish` command?
Found any issues on the last code base?

clumsy-stefan commented:
> Do you use the `publish` command?

no, not until now..

> Found any issues on the last code base?

No, except what I wrote above about the ESP32 build and MQTT (and my opinions to the minimal build)

EDIT: That said, I don't use MQTT at all, so I can't say anything about the whole MQTT stuff...

TD-er committed:
[Build] Fix build issue ESP32

TD-er commented:
> compiling without `USE_MQTT` fails as `src/Commands/MQTT.cpp` should be casted with `#ifdef USES_MQTT` at the top.

Done

> Also my personal view would be to also exclude `WEBSERVER_PINSTATES`, `WEBSERVER_SETUP` and `WEBSERVER_WIFI_SCANNER` in minimal builds...

I will leave `WEBSERVER_SETUP` in the build, since it is also used for initial setup.


> in `define_plugin_sets.h` you check for `WEBSERVER_CUSTOM_BUILD_DEFINED` before actually defining the included parts. Then I think further below on the "minimal OTA Build" part you should also check for this and not just `#undef` a bunch of them...

Good point... done.

TD-er committed:
[Custom Build] Add WEBSERVER_CUSTOM_BUILD_DEFINED to strip out web pages

As suggested by @clumsy-stefan

TD-er committed:
[Custom build] Strip out MQTT commands when no MQTT used.

clumsy-stefan commented:
Cool...

Just one more thing: I still think it just does not make sense to strip out everything (eg. doing a minimal or reduced build) if `MEMORY_ANALYSIS` is defined as suggested in `define_plugin_sets.h` on line 267 (`#if defined(BUILD_MINIMAL_OTA) || defined(MEMORY_ANALYSIS)`. And if it's in there it should be negated (`&& !defined()`)

Especially for analysis purposes you probably want everything in there... Generally I don't see a reason for that check at all... it's just unrelated to what should be included in the build and what not...

TD-er commented:
Problem with the memory analysis is that it may run out of sketch size very easy when running some configurations. 
The memory analysis run is there to test the differences between builds, so we need some base line.

But now we have the new (very cool) inspect tool in PlatformIO, I don't think it is needed anymore.

clumsy-stefan commented:
ah, ok, explained like this it makes sense again... I always thought this is to analyse as much of the memory usage as possible....

TD-er commented:
By the way, not sure if it is related to a change I made or the fact I accidentally uploaded a debug build, but my test node does crash a lot now.
So better not upload to all ;)

clumsy-stefan commented:
my nodes are running for ~3h now with this build with no apparent issues...

TD-er commented:
> my nodes are running for ~3h now with this build with no apparent issues...

The last flash was with the latest committed code.

OK, was apparently related to the "debug build" which is just too slow for actual stable use.

clumsy-stefan commented:
My nodes are just missing the last 3 commits (ESP32, and #define changes).

TD-er committed:
[MQTT] Fix Publish command (#2662)

The parsing was using the wrong indices to determine topic and value.
Fixes: #2663

clumsy-stefan commented:
hmm... size for my 1M builds increased about 8k since yesteday night... reaching nearly the max again... tryingo to find out why...

clumsy-stefan commented:
Ok, there seems to be an issue here again with the defines (similar to what we had with `TASKS_MAX`: if I use the defines in `define_plugin_set.h` and `#undef` all webserver pieces I don't need I get a binary size of  610181 bytes . If I do the same define/undefs in `Custom.h` and define `WEBSERVER_CUSTOM_BUILD_DEFINED` I get a binary size of 612461 bytes. So about 2k difference....

It seems the order of inclusion (or inclusion at all) makes a difference again here...

your view?

TD-er commented:
I guess maybe we have some file which uses the "USES" (or any other define setup in `define_plugin_set.h`) and does not include that .h file.

clumsy-stefan commented:
would it be safe to include `Custom.h` in `define_plugin_sets.h`? Like this it's ensured that the Custom defines are always loaded before the plugin sets.

TD-er commented:
Can you test by adding `#include define_plugin_set.h` in the `ESPEasy_common.h` file?  I think for testing you should have it at the end.

clumsy-stefan commented:
Just tried, still same result, I get 2k more if I do it in "Custom.h" vs. in "define_plugin_sets.h". Also tried to include it at the end of my Custom.h or in ESPEasy_Globals.h. Same result...

Also double-checked that I really do have the same defines/undef's in Custom.h as in define_plugin_sets.h... found no difference..

TD-er commented:
I meant to have the include of the define_plugin_sets.h in the .h file where you also include the Custom.h file.

clumsy-stefan commented:
Tried that (Custom.h gets included ESPEasy-Globals.h and in ESPEasy_common.h) but no change... 

I'll add some #error statements in the parts I exclude, prob. I find what's not excluded

clumsy-stefan commented:
ok, found it, the 
```
    #ifdef USES_SSDP
      #undef USES_SSDP
    #endif
```
part should probably be outside of `WEBSERVER_CUSTOM_BUILD_DEFINED`in `define_plugin_sets.h` (lines 302ff). Or to put it the other ways round: undefining USES_SSDP in Custom.h does not work!

TD-er commented:
Ah, OK, that makes sense I guess :)

clumsy-stefan commented:
yep, question remains, should `USES_SSDP` be user-configurable in `Custom.h` (as currently it's not). It uses about 4k sketch size, which is a lot if not needed....

TD-er committed:
[Build] Fix build issue with core 2.4.1 (F() macro)

TD-er commented:
Would this be a good fix?
```c++
#ifndef USE_CUSTOM_H
    #ifndef USES_SSDP
        #define USES_SSDP
    #endif
#endif
```

TD-er committed:
[Minimize OTA] Allow to define USES_SSDP from Custom.h

clumsy-stefan commented:
>Would this be a good fix?

No, as I thikn the issue is, that it get's not included at the right time!! So the problem is less that it's in `define_plugin_sets.h` but where it's used `Custom.h` is not included..

TD-er commented:
Custom.h is included before the define_plugin_sets.h
Or at least that's how it should be.

clumsy-stefan commented:
yes, but probably not in the files where USES_SSDP is used... I'm just searching...

clumsy-stefan commented:
it seems that `define_plugin_sets.h` gets included in `Networking.ino` however `Custom.h` does not!... So that would explain the issue... I think we need to make sure that `Custom.h` is included everywhere to ensure the defines are present.

clumsy-stefan commented:
In `define_plugin_sets.h`it shoudl be like:
```
// Strip out parts not needed for either MINIMAL_OTA and MEMORY_ANALYSIS
#if defined(BUILD_MINIMAL_OTA) || defined(MEMORY_ANALYSIS)
    #ifndef WEBSERVER_CUSTOM_BUILD_DEFINED
        #ifdef WEBSERVER_TIMINGSTATS
            #undef WEBSERVER_TIMINGSTATS
        #endif
        #ifdef WEBSERVER_SYSVARS
            #undef WEBSERVER_SYSVARS
        #endif
        #ifdef WEBSERVER_NEW_UI
            #undef WEBSERVER_NEW_UI
        #endif
        #ifdef WEBSERVER_I2C_SCANNER
            #undef WEBSERVER_I2C_SCANNER
        #endif
        #ifdef WEBSERVER_FAVICON
            #undef WEBSERVER_FAVICON
        #endif
        #ifdef WEBSERVER_LOG
            #undef WEBSERVER_LOG
        #endif
        #ifdef WEBSERVER_GITHUB_COPY
            #undef WEBSERVER_GITHUB_COPY
        #endif
        #ifdef WEBSERVER_PINSTATES
            #undef WEBSERVER_PINSTATES
        #endif
        #ifdef WEBSERVER_WIFI_SCANNER
            #undef WEBSERVER_WIFI_SCANNER
        #endif
    #endif // WEBSERVER_CUSTOM_BUILD_DEFINED
    #ifdef USES_SSDP
      #undef USES_SSDP
    #endif
#endif
```

And then made sure that `Custom.h` is included everywhere (before `define_plugin_sets.h`) where definitions from either of them are used!

EDIT: i'm mostly offline until late or tomorrow...

TD-er committed:
[Cleanup] Split remaining .h files in Commands dir into .h and .cpp

clumsy-stefan commented:
#4b6619e makes no sense... USES:SSDP should be undefined when doing a minimal build and, if not it should be possible to enable/disable it in Custom .h.

Again, the problem is, that `Custom.h` and `define_plugin_sets.h` does not always get included and therefore the defines are wrong (same as we had with `TASKS_MAX`

For example: even if `USE_CUSTOM_H` is defined, it gets never included when `Networking.ino` is compiled and therefore all defines form Custom.h are not `available`.

This could also be a reason why some users report that setting SSID etc. in `Custom.h` does not work.

clumsy-stefan commented:
the latest commit #e441be1 seem to have this problem on a wider scale. bin size increased again, without changing defines. So it seems, that some of the USES_ and/or WEBSERVER defines don't make it through to the relevant source files...

TD-er commented:
> Again, the problem is, that `Custom.h` and `define_plugin_sets.h` does not always get included and therefore the defines are wrong (same as we had with `TASKS_MAX`

So `Custom.h` should be included by `define_plugin_sets.h`.

`USES_SSDP` is undefined when building MINIMAL_OTA, or at least that was my intent.

clumsy-stefan commented:
>So Custom.h should be included by define_plugin_sets.h.

I think it should included everywhere....

TD-er commented:
> > So Custom.h should be included by define_plugin_sets.h.
> 
> I think it should included everywhere....
If Custom.h is included in define_plugin_sets.h, then it will be included everywhere where the define_plugin_sets.h is included or where the USES_* is used.

clumsy-stefan commented:
but for example not in Network.h I think, even though some defines are used there. 

I just tried to include it at the top of `define_plugin_sets.h` but it doesn't change the outcome...

probably it's just too late... I'll continue tomorrow... but thanks for all hte help!!

TD-er committed:
[Rules] Add some debug logs for checking conditionMatchExtended

TD-er committed:
[clangtidy] Address some warnings reported by CLangTidy

clumsy-stefan reviewed (commented):
Debug statements should be wrapped in `#ifndef BUILD_NO_DEBUG`  (size...)!

TD-er commented:
> Debug statements should be wrapped in `#ifndef BUILD_NO_DEBUG` (size...)!

I don't see the line to which you left a comment.
is it about my (ongoing) debugging of the `conditionMatchExtended` ?

clumsy-stefan commented:
Yes, generally (it's spread over the whole file I think), that debug-logs should be casted, so they can be excluded in minimal builds. Eg. 687, 689ff, 694ff, etc.

TD-er commented:
> Yes, generally (it's spread over the whole file I think), that debug-logs should be casted, so they can be excluded in minimal builds. Eg. 687, 689ff, 694ff, etc.

Do you know, that only now I do realize those markings you make in your posts (with ff) are line numbers and not commit SHA's I could not find?

clumsy-stefan commented:
Oh, sorry for that... I'm kind of used to this notation... What would be the correct way to do it?

btw: minimal build increasd 7k from the commits form 29th to today, without changing any definitions or inclutions, etc....

EDIT: I think I just realised why: `BUILD_NO_DEBUG` does not get defined autmatically anymore for minimal builds (not sure why currently). If I define it myself it's reduced size again...

TD-er commented:
I have been browsing through the binary and not exactly sure yet, but it seems that my small functions to generate html (e.g. `html_TD()`) may be adding more to the binary than simply `TXBuffer += F("<TD>");`
So I will later today run a test build (or you may try it) to have that function as a macro instead.
The new core 2.6.0 does have flash string de-duplication, so now it may be better to use it as a macro.

clumsy-stefan commented:
just edited my comment above... it seems something changed in the logic of the minimal build set in `define_plugin_sets`.... builds are back to what they used to be before yesterday...

TD-er commented:
Just to be sure, you are using core 2.6.0?

clumsy-stefan commented:
yes, 2.6.0 GIT (latest commit, updating regularly)

TD-er committed:
[Cleanup] Change toString(bool) into boolToString(bool)

After finding a number of calls to toString() with only a single parameter, the value would be converted into a bool and then printed.
This will yield in unexpected results without a compiler warning.

TD-er committed:
[Cleanup] Extra boundary checks for deviceIndex

TD-er committed:
[Cleanup] Add boundary checks for accessing UserVar

event->BaseVarIndex is set only at a few places in the code, so adding checks there is enough to make sure all uses of event->BaseVarIndex are safe.

Especially when computing a userVarIndex from event values (user input) the range should be checked before using it to access an element in the UserVar array.

TD-er committed:
[Cleanup] Include define_plugin_sets.h in ESPEasy_common.h and only include common

Just to make sure the include of defaults, Custom.h and the define_plugin_sets.h is done in the right order and always in that order.

TD-er committed:
[Build] Fix build error in P086_Homie 

One closing curly brace too many.

TD-er commented:
@clumsy-stefan Have you seen the latest commits I added to this PR?
I think it should mainly address the issues you had with ArduinoIDE.

The `ESPEasy-common.h` does include
- `Custom.h`
- `define_plugin_sets.h`

So these will be bound to eachother and always include the other if one is included (and in specific order)

clumsy-stefan commented:
yes saw this one, but can't see any differnec ein building.

Two remoarks:
1) I think the `#ifndef USE_CUSTOM_H` in `define_plugin_sets.h` (from line 109) is not needed.

2) lines 268 and folowing should be as in comment https://github.com/letscontrolit/ESPEasy/pull/2698#issuecomment-547945642 (`USE_SSDP` outside of `WEBSERVER_CUSTOM_BUILD_DEFINED`

other than this everything seems fine to me..

TD-er commented:
> * I think the `#ifndef USE_CUSTOM_H` in `define_plugin_sets.h` (from line 109) is not needed.

It is needed, since the file `Custom.h` may not exist and then the #include will fail the build.

clumsy-stefan commented:
lines 109 and following in `define_plugin_sets.h`:
```
#ifndef USE_CUSTOM_H
    #ifndef USES_SSDP
        #define USES_SSDP
    #endif
#endif
```

Nothing is included here??

TD-er commented:
Oh that one.
Well, it does add the define, unless it is overriden in Custom.h.

clumsy-stefan commented:
yes, but like this, if there is a `Custom.h` and ÙSES_SSDP` is not defined in there, it won't get defined at all... Which is (I think) the opposite of wht has been default until now.

important is, that it get's striped out on minimal builds...

TD-er commented:
Well the problem is that there is no "don't want it" option for it, which makes it a bit different.

clumsy-stefan commented:
agree on that, so I can live with this...

what also seems to have changed is, that only when `PLUGIN_BUILD_MINIMAL_OTA`is defined `BUILD_MINIMAL_OTA`gets defined. I somehow feel it should be the other way round? But then, the longer I try to understand the logic of `define_plugin_sets.h` together with `Custiom.h`the more confused I get ;) So you have my full support, that the logic here is as intended...

TD-er commented:
Does make sense that the other way around may sound more logical
I don't think it is intended like this, but may have grown into this it is now.

TD-er commented:
I will merge it now, if you think it is all running fine?

clumsy-stefan commented:
I did not test exhaustingly, but from what I have seen, my builds ran stable for quite some time now...

So keep fingers crossed and hope that's true for the standard builds and everyone else!

TD-er commented:
Well, apart from the ones I added yesterday, the other commits have been tested for a while by several people.
So it will be better than it was :)
